### ISSUE TITLE
Sink subprocesses not exiting

### ISSUE THREAD

**Issue by MrDrMcCoy** [statsite/statsite#259]
## Issue description
I'm having a strange issue: On certain systems, the sink scripts in their subshells never close, and their processes build up in a manner that would eventually crash the system. On other systems in our environment, everything runs as intended. The sinks that never die do send metrics as expected, but they seemingly never get to the part where they perform their final task and exit. 

We are using a custom Python sink to help maintain compatibility with the StatsD that we are replacing (Statsite has just different enough behavior to break our dashboards). While this seems like an easy culprit for the issue, the fact that it works in QA and not Prod indicates otherwise. Stranger still, implementing a 20sec timeout within the sink script, and having Statsite call it with `timeout -s 9 30 python sink.py` also fails to cause the sink to exit and leaves the processes to pile up. My sink times out as expected when flooding it with sample sinks (generated by another Python script), and exits cleanly before the timeout period in our QA environment when called by Statsite.

Here are a few details about the machines:

#### QA (working)
KVM VM
CPU: 2 cores
RAM: 16GB
OS: Ubuntu 14.04
Python version: 2.7.6
Libc version: Ubuntu EGLIBC 2.19-0ubuntu6.11
GCC version: Ubuntu 4.8.4-2ubuntu1~14.04.3
Metrics every 10 sec: ~400,000

#### Prod (dangling subshells)
Bare metal
CPU: 2x E5-2690v3 (16 cores total + Hyperthreading)
RAM: 128GB
OS: Ubuntu 14.04
Python version: 2.7.6
Libc version: Ubuntu EGLIBC 2.19-0ubuntu6.7
GCC version: Ubuntu 4.8.4-2ubuntu1~14.04.1
Metrics every 10 sec: ~40,000

#### Deb package build box
Vagrant image
OS: Ubuntu 14.04
Libc version: Ubuntu EGLIBC 2.19-0ubuntu6.11
GCC version: Ubuntu 4.8.4-2ubuntu1~14.04.3
Statsite version: git commit version 1dded468f

**Here is the sink I wrote** (its fugly but hopefully easy to read, sorry):
https://gist.github.com/MrDrMcCoy/793381bf0480f4487d4184285a4e1d3f

**And our Statsite config** (destination Graphite host is filled out by Ansible):
https://gist.github.com/MrDrMcCoy/3e301f42971acb0d5218bf15cd3c2d41

Is there anything obvious that you can think of that could cause the sink subshells to hang? Is there enough of a difference in our Libc versions to cause a bug? Is there a bug somewhere else in my code or yours? We would really appreciate any and all help.

**Comment by johnkeates** [comment]
I'm not sure about the influence a slight difference in EGLIBC would have, but if any, it would be in excl().

Statsite basically takes stream_cmd, and uses it in streaming.c: res = execl("/bin/sh", "sh", "-c", cmd, NULL); where cmd is stream_cmd.

So basically, you get sh -c $stream_cmd and that just goes on until stream_cmd exits by itself. There are unit tests that do stuff like cat and tee to check and confirm the functionality. I believe the issue you're experiencing might be related to the sink itself, but I'm no python expert so I can't inspect yours (or rather: I don't see something that's obviously wrong or anything like that).

**Comment by MrDrMcCoy** [comment]
Just to try and make sure i'm not doing something inefficient in how my Python code handles the input, I changed this:

```
for line in sys.stdin:
    process_sink(line)
end()
```

to 

```
input_lines = []
while True:
    try:
        input_lines.append(raw_input().strip())
    except EOFError:
        break
for line in input_lines:
    if line != '':
        process_sink(line)
end()
```

Which is more or less what the provided sink script does. Sadly, it exhibits the same behavior of working just fine in QA, but leaves dangling processes in Prod :(

**Comment by johnkeates** [comment]
Since it's an environmental difference, is it perhaps configuration of forking/spawning/exec'ing? Is there a difference in AppArmor, SELinux or sysctl? At this point it seems more of a change beween how the shell is supposed to clean up after itself. Maybe sh points to a different interpreter?

**Comment by MrDrMcCoy** [comment]
I wasn't able to find any interesting differences between the environments in QA and prod. However, I was able to cause the script to indefinitely hang artificially: 

```
~# until false ; do echo 'stuff' ; while true; do dd if=/dev/null 2>/dev/null ; done ; done | sh -c '/usr/bin/timeout -s 9 30 /usr/bin/python /usr/share/statsite/sinks/statsite2graphite.py'
```

It seems that a pipe that starts with data, but never closes or otherwise pipes null prevents `/usr/bin/timeout` and the Python timeout function within the script from ever trying to kill anything.  Can you think of any reason that the fork would leave its pipe open in this manner?

**Comment by johnkeates** [comment]
What does sh point to on your distro? It seems to me that it's something the shell is supposed to handle. You could try an strace to see if maybe it's getting the proper signals but not handling it right

**Comment by MrDrMcCoy** [comment]
On Ubuntu, `sh` is provided by [dash](https://packages.ubuntu.com/trusty/amd64/dash/filelist).

I was able to simplify the test case and get some interesting-looking `strace` output, but I'm not sure how to interpret it:

```
~# while true ; do true ; done | strace sh -c '/usr/bin/timeout -s 9 10 cat'
execve("/bin/sh", ["sh", "-c", "/usr/bin/timeout -s 9 10 cat"], [/* 24 vars */]) = 0
brk(0)                                  = 0x7faf48fbf000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7faf47f2b000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=42963, ...}) = 0
mmap(NULL, 42963, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7faf47f20000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0P \2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1853216, ...}) = 0
mmap(NULL, 3961536, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7faf47725000
mprotect(0x7faf478e3000, 2093056, PROT_NONE) = 0
mmap(0x7faf47ae2000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1bd000) = 0x7faf47ae2000
mmap(0x7faf47ae8000, 17088, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7faf47ae8000
close(3)                                = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7faf47f1f000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7faf47f1d000
arch_prctl(ARCH_SET_FS, 0x7faf47f1d740) = 0
mprotect(0x7faf47ae2000, 16384, PROT_READ) = 0
mprotect(0x7faf47f2d000, 8192, PROT_READ) = 0
mprotect(0x7faf47d0f000, 4096, PROT_READ) = 0
munmap(0x7faf47f20000, 42963)           = 0
getpid()                                = 6501
rt_sigaction(SIGCHLD, {0x7faf47d24460, ~[RTMIN RT_1], SA_RESTORER, 0x7faf4775bcb0}, NULL, 8) = 0
geteuid()                               = 0
brk(0)                                  = 0x7faf48fbf000
brk(0x7faf48fe0000)                     = 0x7faf48fe0000
getppid()                               = 6498
stat("/root", {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
stat(".", {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
rt_sigaction(SIGINT, NULL, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGINT, {0x7faf47d24460, ~[RTMIN RT_1], SA_RESTORER, 0x7faf4775bcb0}, NULL, 8) = 0
rt_sigaction(SIGQUIT, NULL, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGQUIT, {SIG_DFL, ~[RTMIN RT_1], SA_RESTORER, 0x7faf4775bcb0}, NULL, 8) = 0
rt_sigaction(SIGTERM, NULL, {SIG_DFL, [], 0}, 8) = 0
rt_sigaction(SIGTERM, {SIG_DFL, ~[RTMIN RT_1], SA_RESTORER, 0x7faf4775bcb0}, NULL, 8) = 0
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7faf47f1da10) = 6502
wait4(-1, [{WIFSIGNALED(s) && WTERMSIG(s) == SIGKILL}], 0, NULL) = 6502
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=6502, si_status=SIGKILL, si_utime=0, si_stime=0} ---
rt_sigreturn()                          = 6502
write(2, "Killed\n", 7Killed
)                 = 7
exit_group(137)                         = ?
+++ exited with 137 +++
^C
```

After `timeout` prints killed and `strace` prints the exited message, the process remains running until I send it a CTRL+C. I was able to get similar behavior with `bash`. I'm unsure what to make of this.

**Comment by MrDrMcCoy** [comment]
Just for fun, I reproduced this behavior on Ubuntu 12.04, 14.04 and 16.04; CentOS 7, and OpenSUSE Tumbleweed. Are we sure that it isn't the parent that is supposed to close the pipe?

**Comment by johnkeates** [comment]
Well, the parent closes the fd (file descriptor) which then closes the pipe as a result as far as I know. It's how all the other sinks work that are used by default. Do they hang for you as well? I checked some of my systems, most of them point to dash as well, so that's probably not an issue.

**Comment by MrDrMcCoy** [comment]
I added a bit of extra logging and assurance that the pipe is getting closed properly:
```
input_lines = []
for line in sys.stdin.read().splitlines():
    input_lines.append(line)
sys.stdin.close()
logger.info('Finished receiving flush, processing ' + str(len(input_lines)) + ' metrics...')
for line in input_lines:
    if line != '':
        process_sink(line)
end()
```
Its logging indicated that stdin is indeed getting closed.

Also just to try and rule out a [race condition](https://stackoverflow.com/questions/38059954/closing-pipe-does-not-interrupt-read-in-child-process-spawned-from-thread), I increased the sink interval to 30, and dropped the `timeout -s 9` to 25 and the timeout within the script to 20. However, the processes continued to pile up.

As a final test, I removed the `timeout` command and the timeout function within the script, jacked up the logging, and watched it all run. It worked. My theory is that the addition of another subshell (as happens with `timeout`) prevents the proper destruction of the pipes and leaves the process dangling. I don't think that there is a problem with statsite/libc/dash/python after all, just an edge case resulting from making things too complicated. Sorry to have troubled you a second time.
